[juicenet] Initial contribution

jsjames opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Skinah commented on bundles/org.openhab.binding.juicenet/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.1.0-SNAPSHOT</version>
```
```suggestion
    <version>3.2.0-SNAPSHOT</version>
```
needed to get it compiling now the project has released 3.1 Stable

Skinah commented on bundles/pom.xml:
```diff
@@ -1,597 +1,1145 @@
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    
   <modelVersion>4.0.0</modelVersion>
-
+    
```
Not sure what happened but this file is full of changes. Probably easiest to replace the file from the project and make the needed changes to the fresh file.

Skinah commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+		<item-type>Number:Length</item-type>
+		<label>Mileage Range</label>
+		<description>Car mileage range.</description>
+		<state readOnly="true" pattern="%,d miles"/>
```
Why is the pattern hard coded to display miles?

jsjames replied:
This one can be made to use UOM since there is a distance UOM defined.

jlaur replied:
@Skinah - I'm resolving this comment on your behalf since this was fixed recently.

Skinah commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+		<state readOnly="true" pattern="$%.2f"/>
+	</channel-type>
+
+	<channel-type id="whpermile_type">
+		<item-type>Number:Power</item-type>
+		<label>Watts per mile</label>
+		<description>Watts per mile.</description>
```
Have you implemented UOM? See https://www.openhab.org/docs/concepts/units-of-measurement.html#defining-channeltypes

Not every country uses Miles so it should be possible to use UOM so it automatically changes depending on which country the openHAB is setup to be in.

jsjames replied:
This one and others, such as miles per gallon, cannot be done without creating new UOM definitions on openhab_core.

jlaur reviewed (changes requested):
Updating PR status.

lolodomo commented:
@jsjames : let us know if you want to finish this binding.

jsjames commented:
Hi - there are 2 CI / Build checks failing on this PR which have to do with karaf issues.  are these general build issues unrelated to this PR?

wborn commented on bundles/org.openhab.binding.juicenet/pom.xml:
```diff
@@ -0,0 +1,16 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.3.0-SNAPSHOT</version>
```
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/JuiceNetBindingConstants.java:
```diff
@@ -0,0 +1,76 @@
+    private static final String BINDING_ID = "juicenet";
+
+    // List of Bridge Type
+    public static final String BRIDGE = "juicenet-account";
```
"juicenet-" prefix can be removed.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/JuiceNetBindingConstants.java:
```diff
@@ -0,0 +1,76 @@
+    public static final String BRIDGE = "juicenet-account";
+
+    // List of all Device Types
+    public static final String DEVICE = "juicenet-device";
```
"juicenet-" prefix can be removed.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,293 @@
+
+    protected String apiToken = "";
+    protected JuiceNetHttp httpApi = new JuiceNetHttp();
+    @Nullable
```
What's the reason for this to be nullable?

jsjames replied:
Since the class is @NonNullByDefault, it would require bridgeUID to be set in the constructor (there is none for JuiceNetApi). This value is only set in the initialize() function when it is passed as a parameter.

jlaur replied:
Correct, I must have misread something when posting that comment.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,293 @@
+        HttpResponse<String> response;
+
+        params.put("cmd", "get_account_units");
+        params.put("device_id", bridgeUID.getAsString()); // bridgeUID.toString());
```
Can the comment be omitted?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApiException.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     *
+     */
+    private static final long serialVersionUID = 1L;
```
Please generate unique UID to ensure correct serialization.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetHttp.java:
```diff
@@ -0,0 +1,95 @@
+
+    public HttpResponse<String> httpGet(String url, @Nullable String paramString)
+            throws IOException, InterruptedException {
+        HttpRequest request = HttpRequest.newBuilder(URI.create(url + paramString)).GET().build();
```
Could the frameworks' shared Jetty HTTP client be used here instead?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,293 @@
+    @Nullable
+    protected ThingUID bridgeUID;
+
+    public boolean initialize(String apiToken, ThingUID bridgeUID) throws JuiceNetApiException {
```
This method could be `void` as only `true` will ever be returned.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,158 @@
+            try {
+                listDevices = api.queryDeviceList();
+            } catch (JuiceNetApiException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, e.toString());
```
This exception is probably thrown in case of network/API issues? In this case it should probably be **COMMUNICATION_ERROR**, not **CONFIGURATION_ERROR**.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,158 @@
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduler.execute(() -> {
```
Since this is a one-time shot that will return in case of network issues, this will most likely result in the binding never being able to initialize and get back online in such situations?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,158 @@
+            pollingJob = scheduler.scheduleWithFixedDelay(this::pollDevices, 60, config.refreshInterval,
+                    TimeUnit.SECONDS);
+
+            logger.trace("Bridge is ONLINE");
```
This logging can be omitted since the status change will be logged by the framework.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+
+    // This holds the last time the info was updated and is compared everytime we get device status to determine
+    // if the deviceInfo should be updated. Set the time in the past to ensure info is updated on initial call.
+    protected ZonedDateTime infoTimestamp = ZonedDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC);
```
`Instant` would be more appropriate here since time zone is not used.

jsjames replied:
This must have been a left-over and is not used in the code.  It has been deleted.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Example for background initialization:
```
Can this comment be removed?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Example for background initialization:
+        scheduler.execute(() -> {
```
Since this is a one-time shot that will return in case of network issues, this will most likely result in the binding never being able to initialize and get back online in such situations?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+            try {
+                queryDeviceStatusAndInfo();
+            } catch (JuiceNetApiException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, e.toString());
```
This exception is probably thrown in case of network/API issues? In this case it should probably be **COMMUNICATION_ERROR**, not **CONFIGURATION_ERROR**.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("handleCommand");
```
If this is really needed, I'd propose TRACE log level and to include channelUID and command.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,295 @@
+        Instant instant = Instant.ofEpochSecond(localEpochSeconds);
+        ZonedDateTime zdt = instant.atZone(ZoneId.of("UTC"));
+
+        return zdt.withZoneSameLocal(ZoneId.systemDefault());
```
If this ZonedDateTime is exposed to channels, please use `TimeZoneProvider` for getting configured time zone.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/i18n/juicenet.properties:
```diff
@@ -0,0 +1,21 @@
+# FIXME: please substitute the xx with a proper locale, ie. de
+# FIXME: please do not add the file to the repo if you add or change no content
+# binding
```
Please remove these lines.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/i18n/juicenet.properties:
```diff
@@ -0,0 +1,21 @@
+binding.juicenet.name = JuiceNet Binding
+binding.juicenet.description = This binding interfaces to a JuiceNet EV Charger. 
+
+# thing types
+thing-type.juicenet.sample.label = <Your localized Thing label>
+thing-type.juicenet.sample.description = <Your localized Thing description>
+
+# thing type config description
+thing-type.config.juicenet.sample.hostname.label = <Your localized config parameter label>
+thing-type.config.juicenet.sample.hostname.description = <Your localized config parameter description>
+thing-type.config.juicenet.sample.password.label = <Your localized config parameter label>
+thing-type.config.juicenet.sample.password.description = <Your localized config parameter description>
+thing-type.config.juicenet.sample.refreshInterval.label = <Your localized config parameter label>
+thing-type.config.juicenet.sample.refreshInterval.description = <Your localized config parameter description>
+
+# channel types
+channel-type.juicenet.sample-channel.label = <Your localized Channel label>
+channel-type.juicenet.sample-channel.description = <Your localized Channel description>
```
Please remove all examples and instructions, and regenerate file: https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-account.xml:
```diff
@@ -0,0 +1,23 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<bridge-type id="juicenet-account">
```
Prefix can be omitted:
```suggestion
	<bridge-type id="account">
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="juicenet-device">
```
Prefix can be omitted:
```suggestion
	<thing-type id="device">
```

jlaur reviewed (changes requested):
Thanks for your contribution! I realize this PR has been around for some time, sorry for the delay. If you are still interested in a review, please let me know. I have provided some initial feedback, but can go into a full review if you'd like.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,330 @@
+    }
+
+    protected ZonedDateTime toZonedDateTime(long localEpochSeconds) {
+        Instant instant = Instant.ofEpochSecond(localEpochSeconds);
+        ZonedDateTime zdt = instant.atZone(ZoneId.of("UTC"));
+
+        Bridge bridge = Objects.requireNonNull(getBridge());
+        JuiceNetBridgeHandler handler = (JuiceNetBridgeHandler) Objects.requireNonNull(bridge.getHandler());
+
+        return zdt.withZoneSameLocal(handler.getTimeZoneProvider().getTimeZone());
```
This could be simplified by having `TimeZoneProvider` stored in the handler instance:
```suggestion
        return Instant.ofEpochSecond(localEpochSeconds).atZone(timeZoneProvider.getTimeZone());
```

jsjames replied:
I originally had it implemented that way, but then thought it would be better to store only once in the bridge vs. each device.  I agree though, this is simpler.

jlaur replied:
Leaving this comment open, since only half of the proposal was adopted. :-) You do not need the temporary `ZonedDateTime` in UTC.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-account.xml:
```diff
@@ -0,0 +1,23 @@
+
+	<bridge-type id="account">
+		<label>JuiceNet Account</label>
+		<description>JuiceNet account</description>
```
Could the description be extended for being a bit more descriptive than the label?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-account.xml:
```diff
@@ -0,0 +1,23 @@
+			</parameter>
+			<parameter name="refreshInterval" type="integer" unit="s" min="60">
+				<label>Refresh Interval</label>
+				<description>Interval the device is polled in sec.</description>
```
```suggestion
				<description>Interval the device is polled in seconds.</description>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+
+	<thing-type id="device">
+		<supported-bridge-type-refs>
+
```
This linefeed can be removed.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="chargingStateType">
+		<item-type>String</item-type>
+		<label>Charging State</label>
+		<description>The charing state (Start Charging, Smart Charging, Stop Charging).</description>
```
```suggestion
		<description>The charging state (Start Charging, Smart Charging, Stop Charging).</description>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="targetTimeType">
+		<item-type>DateTime</item-type>
+		<label>Target Time</label>
+		<description>“Start charging” start time, or time to start when overriding smart charging.</description>
```
Icon could be added:
```suggestion
		<description>“Start charging” start time, or time to start when overriding smart charging.</description>
		<category>Time</category>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="plugUnplugTimeType">
+		<item-type>DateTime</item-type>
+		<label>Plug/Unplug Time</label>
+		<description>Last time of either plug-in or plug-out.</description>
```
Icon could be added:
```suggestion
		<description>Last time of either plug-in or plug-out.</description>
		<category>Time</category>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="unitTimeType">
+		<item-type>DateTime</item-type>
+		<label>Unit Time</label>
+		<description>Current time on the unit.</description>
```
Icon could be added:
```suggestion
		<description>Current time on the unit.</description>
		<category>Time</category>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="temperatureType">
+		<item-type>Number:Temperature</item-type>
+		<label>Temperature</label>
+		<description>Current temperature at the unit.</description>
```
Icon could be added:
```suggestion
		<description>Current temperature at the unit.</description>
		<category>Temperature</category>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="wattPowerType">
+		<item-type>Number:Power</item-type>
+		<label>Power</label>
+		<description>Current charging power. (W)</description>
+		<state readOnly="true" pattern="%.2f kW"/>
```
There is some inconsistency between description and supplied unit (W vs kW). Best would probably be to drop "W" from the description and replace "kW" with "%unit%" as unit will be supplied when creating the `QuantityType`:
```suggestion
		<description>Current charging power.</description>
		<state readOnly="true" pattern="%.2f %unit%"/>
```

Additionally, reconsider the name of the channel type. Perhaps "chargingPower"?

Some similar comments about unit for other channel type definitions, I will go through them later.

jsjames replied:
That makes sense.  I was torn between adopting the names used in the JuiceNet API vs. coming up with better names.   Related, there are several quantityTypes which don't exists (such as EnergyPerMile o ro miles per gallon).  Those i have just kept as Decimal values.

jlaur replied:
Regarding names used in the API, don't feel bound by them. This binding is an abstraction from that API, so that users can have a unified experience interacting with various equipment implementing completely different API's.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="secondsChargingType">
+		<item-type>Number:Time</item-type>
+		<label>Charging Time</label>
+		<description>Charging time since plug-in time. (s)</description>
```
Icon could be added:
```suggestion
		<description>Charging time since plug-in time. (s)</description>
		<category>Time</category>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+	<channel-type id="ecostType">
+		<item-type>Number</item-type>
+		<label>Utility Energy Cost</label>
+		<description>Cost of electricity from utility company. ($/kWh)</description>
```
Is dollars per kWh the concrete unit, or will local currency per kWh be used?

jsjames replied:
The API documentation I have is very US centric and hard-codes things like Miles.  I presume though that this currency could be any currency.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+		<description>JuiceBox EV-Charger</description>
+
+		<channels>
+			<channel id="name" typeId="nameType"/>
```
General: The channel type definitions can have the same name as the channels, so the "Type" postfix is not needed.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,271 @@
+
+	<channel-type id="whPerMileType">
+		<item-type>Number:Power</item-type>
+		<label>Watts per mile</label>
```
Seems like it should be:
```suggestion
		<label>Watt Hours Per Mile</label>
```

I'm wondering if this could somehow be made more generic or customizable. It's energy/distance, and for the distance part most European users will probably prefer an SI unit, i.e. km.

jsjames replied:
I looked at this - but I believe we'd have to add Energy / Distance as a unit for QuantityTypes, correct?  Or, do you see another way to do this?

jlaur replied:
I think I need some help here, @openhab/add-ons-maintainers. This would require a new unit like W/m, right?
For reference: https://www.openhab.org/docs/concepts/units-of-measurement.html#list-of-units

jlaur reviewed (changes requested):
I have posted some additional comments now that you are progressing with the PR. As promised, I will do a full review, but didn't make it yet.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/JuiceNetHandlerFactory.java:
```diff
@@ -0,0 +1,73 @@
+    @Activate
+    public JuiceNetHandlerFactory(@Reference HttpClientFactory httpClientFactory,
+            @Reference TimeZoneProvider timeZoneProvider) {
+        this.httpClientFactory = httpClientFactory;
```
Sounds good.  I didn't know if the "common client" would ever change from the factory during runtime so I did it this way.  I'll update this.

jsjames commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,218 @@
+
+| channel           | type              | read-only | description                  |
+|----------         |--------           |------------------------------|
+| charging_state    | String            | N         | Current charging state (Start Charging, Smart Charging, Stop Charging). |
```
sorry, missed that.

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,214 @@
+| chargingState     | String            | N         | Current charging state (Start Charging, Smart Charging, Stop Charging). |
+| state             | String            | Y         | This is the current device state (Available, Plugged-In, Charging, Error, Disconnected).  |
+| override          | Switch            | Y         | Smart charging is overridden. |
+| chargingTimeL eft | Number:Time       | Y         | Charging time left (seconds). |
```
```suggestion
| chargingTimeLeft  | Number:Time       | Y         | Charging time left (seconds). |
```

jlaur replied:
Minor formatting problem now.

jlaur reviewed (changes requested):
I had some unpublished comment from previous iteration. Will flush the now as I see a few are already resolved. I will continue from last commit.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,196 @@
+        try {
+            api.initialize(config.apiToken, this.getThing().getUID(), this.httpClientFactory);
+        } catch (JuiceNetApiException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, e.toString());
```
api.initialize was wrongly tagged as throwing an exception.  I just have changed the code and forgot to remove.  I removed the try/catch in this section.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,276 @@
+		<item-type>Number:ElectricPotential</item-type>
+		<label>Voltage</label>
+		<description>Current voltage. (V)</description>
+		<state readOnly="true" pattern="%d V"/>
```
Unit is given by QuantityType creation, right?
```suggestion
		<state readOnly="true" pattern="%d %unit%"/>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,276 @@
+		<item-type>Number:ElectricCurrent</item-type>
+		<label>Current</label>
+		<description>Current charging current. (A)</description>
+		<state readOnly="true" pattern="%.1f A"/>
```
Unit is given by QuantityType creation, right?
```suggestion
		<state readOnly="true" pattern="%.1f %unit%"/>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,276 @@
+		<item-type>Number:ElectricCurrent</item-type>
+		<label>Current Limit</label>
+		<description>Max charging current allowed. (A)</description>
+		<state pattern="%d A"/>
```
Unit is given by QuantityType creation, right?
```suggestion
		<state pattern="%d %unit%"/>
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,276 @@
+	<channel-type id="carBatterySize">
+		<item-type>Number:Energy</item-type>
+		<label>Car Battery Pack Size</label>
+		<description>Car battery pack size. (Wh)</description>
```
Unit doesn't matter actual unit:
```suggestion
		<description>Car battery pack size.</description>
```

jlaur reviewed (changes requested):
Once again I ran out of time, sorry. I'll need to resume later, yet I will post my next batch of feedback, giving you a chance to either wait for the rest or go through them now if you wish.

I believe there are also some compiler warnings to be checked.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/JuiceNetBindingConstants.java:
```diff
@@ -0,0 +1,76 @@
+    public static final String DEVICE_ENERGY = "energy";
+    public static final String DEVICE_SAVINGS = "savings";
+    public static final String DEVICE_POWER = "power";
+    public static final String DEVICE_CHARGING_TIME = "secondsCharging";
```
I changed this to chargingTime.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,327 @@
+    @Override
+    public void initialize() {
+        logger.trace("JuiceNetDeviceHandler:initialize");
+        Map<String, String> properties = editProperties();
```
that is done in the discovery service (notifyDiscoveryDevice) when the device is created.

jlaur replied:
Okay, two comments then:
- Why is `editProperties()` called here?
- How will it work for manually created things?

jsjames replied:
From what I can see, there isn't a "getProperties()". Is there a better way to get the properties?

Yeah, that's a hole as I never tested manual configuration of the binding.  Let me implement that and checkin.

jsjames replied:
When looking at available methods for the ThingHandler, i only saw editProperties().  I see that the Thing has a getProperties() method.  Seems odd to have this split between the Thing and Thing Handler this way??

jlaur replied:
Yeah, `getThing().getConfiguration().getProperties()`. Not sure why.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,276 @@
+		<description>JuiceBox EV-Charger</description>
+
+		<channels>
+			<channel id="name" typeId="name"/>
```
There is a property with the same name.  However, this is also used in the widget and can be linked to an item.  I didn't see a way to link an property to the item, correct?

jlaur replied:
Correct.

jsjames commented:
> 

Thanks for all the time you are dedicating to this.  I think there are a few compiler warnings regarding potential null pointer access, but those should all be clean by design.  I dont' see any other compiler issues on my end.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+
+    public List<JuiceNetApiDevice> queryDeviceList()
+            throws JuiceNetApiException, IOException, InterruptedException, TimeoutException, ExecutionException {
+        Map<String, Object> params = new HashMap<>();
+        ContentResponse response;
+
+        params.put("cmd", "get_account_units");
+        params.put("device_id", bridgeUID.getAsString());
+        params.put("account_token", apiToken);
+
+        response = httpApi.httpPost(API_ACCOUNT, params);
```
After switching to Jetty, this seems overly complicated and could be simplified into something like:
```suggestion
        Request request = httpClient.newRequest(API_ACCOUNT);
        request.param("cmd", "get_account_units");
        request.param("device_id", bridgeUID.getAsString());
        request.param("account_token", apiToken);
        ContentResponse response = request.send();
```

which would eliminate the need for `JuiceNetHttp`. This would also give you more control for extending with custom headers, etc.

jsjames replied:
request.param is only for parameters on the URI.  This API requires the "parameters" to be sent as json as part of the body.  I'll still attempt to simplify, but likely still need some Map and then us GSON to create the json body.

jlaur replied:
Sorry for misleading, I had mistaken `paramsToBody` for `paramsToUrl` where my comment would apply to the latter, i.e. the `httpGet` variant.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+
+        logger.trace("{}", response.getContentAsString());
+
+        JsonObject jsonResponse = JsonParser.parseString(response.getContentAsString()).getAsJsonObject();
```
This can throw `JsonParseException` if content is not valid JSON.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+                    + response.getStatus() + ").");
+        }
+
+        logger.trace("{}", response.getContentAsString());
```
Some overhead could be avoided by storing content in local String as `getContentAsString()` as called again in next line.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+        logger.trace("{}", response.getContentAsString());
+
+        JsonObject jsonResponse = JsonParser.parseString(response.getContentAsString()).getAsJsonObject();
+        boolean success = jsonResponse.get("success").getAsBoolean();
```
`NullPointerException` possible.

jlaur replied:
Code was moved, but still NPE possible if member "success" is not found in the response.

jlaur replied:
@jsjames - I was checking for any open comments and found this one which seems to have slipped.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    public JuiceNetApiInfo queryInfo(String token)
+            throws IOException, InterruptedException, JuiceNetApiException, TimeoutException, ExecutionException {
```
Same comment here as for `queryDeviceList`.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+            throw new JuiceNetApiException("getDeviceStatus from JuiceNet failed, please check configuration.");
+        }
+
+        final JuiceNetApiDeviceStatus deviceStatus = new Gson().fromJson(jsonResponse, JuiceNetApiDeviceStatus.class);
```
`JsonSyntaxException` possible.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+        params.put("device_id", bridgeUID.getAsString());
+        params.put("token", token);
+
+        response = httpApi.httpPost(API_DEVICE, params);
```
Same comment here as for `queryDeviceList`. POST request can be set up with `request.method(HttpMethod.POST);`

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+                    + response.getStatus() + ").");
+        }
+
+        logger.trace("{}", response.getContentAsString());
```
Some overhead could be avoided by storing content in local String as `getContentAsString()` as called again in next line.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    public JuiceNetApiTouSchedule queryTOUSchedule(String token)
+            throws IOException, InterruptedException, JuiceNetApiException, TimeoutException, ExecutionException {
```
Same comment here as for `queryDeviceList`.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    public void setOverride(String token, int energy_at_plugin, Long override_time, int energy_to_add)
+            throws IOException, InterruptedException, JuiceNetApiException, TimeoutException, ExecutionException {
```
Same comment here as for `queryDeviceList`.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/config/JuiceNetBridgeConfiguration.java:
```diff
@@ -0,0 +1,29 @@
+     * Sample configuration parameters. Replace with your own.
+     */
+    public String apiToken = "";
+    public int refreshInterval;
```
Default should be set here also:
```suggestion
    public int refreshInterval = 60;
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/config/JuiceNetBridgeConfiguration.java:
```diff
@@ -0,0 +1,29 @@
+ */
+@NonNullByDefault
+public class JuiceNetBridgeConfiguration {
+    /**
+     * Sample configuration parameters. Replace with your own.
+     */
```
Can this comment be removed?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/dto/JuiceNetApiCar.java:
```diff
@@ -0,0 +1,30 @@
+ */
+@NonNullByDefault
+public class JuiceNetApiCar {
+    public int car_id;
```
To fix this and all other DTO members not following naming convention:
```suggestion
    @SerializedName("car_id")
    public int carID;
```

See checkstyle report in `target/report.html` after `mvn clean install`.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+
+    private final Logger logger = LoggerFactory.getLogger(JuiceNetBridgeHandler.class);
+
+    protected JuiceNetBridgeConfiguration config = new JuiceNetBridgeConfiguration();
```
```suggestion
    private JuiceNetBridgeConfiguration config = new JuiceNetBridgeConfiguration();
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+        return api;
+    }
+
+    protected @Nullable ScheduledFuture<?> pollingJob;
```
```suggestion
    private @Nullable ScheduledFuture<?> pollingJob;
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    protected @Nullable ScheduledFuture<?> pollingJob;
+    protected List<JuiceNetApiDevice> listDevices = Collections.<JuiceNetApiDevice> emptyList();
```
This seems used only locally in `goOnline`, so can be removed as instance variable.

jsjames replied:
Correct, however, I am fixing the issue where the binding didn't support manual configuration and need to store the info here since the token is required for the device.  Once I get this fix checked in, let's review this one again.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+
+    protected @Nullable ScheduledFuture<?> pollingJob;
+    protected List<JuiceNetApiDevice> listDevices = Collections.<JuiceNetApiDevice> emptyList();
+    protected @Nullable JuiceNetDiscoveryService discoveryService;
```
```suggestion
    private @Nullable JuiceNetDiscoveryService discoveryService;
```

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+# JuiceNet Binding
+
+The JuiceNet binding will interface with the cloud portal to get status and manage your JuiceBox EV charger(s).
+In addition to getting the status of various items from the ev-charger, it is also possible to start and stop charging sessions.
```
For being consistent with first line which I believe is correct:
```suggestion
In addition to getting the status of various items from the EV charger, it is also possible to start and stop charging sessions.
```

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+
+## Thing Configuration
+
+The only configuration required is to create a JuiceNet account thing and fill in the appropriate API Token.
```
```suggestion
The only configuration required is to create a JuiceNet account thing and fill in the appropriate API token.
```

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+The only configuration required is to create a JuiceNet account thing and fill in the appropriate API Token.
+The API token can be found on the Account page at https://home.juice.net/Manage.
+
+Once, the JuiceNet Account thing has been created, all JuiceBox EV-chargers associated with that account will be discovered and added to the inbox.
```
```suggestion
Once the JuiceNet Account thing has been created, all JuiceBox EV chargers associated with that account will be discovered and added to the inbox.
```

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+## Channels
+
+| channel           | type              | read-only | description                  |
+|----------         |--------           |------------------------------|
```
This is not correctly formatted and will be wrongly rendered.

jsjames replied:
Thanks - have you been able to get tables to correctly render in the Eclipse Preview?

jlaur replied:
I haven't tried that, sorry. I'm using the viewer in Directory Opus sometimes, as well as the GitHub viewer when changes are pushed.

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+| lifetimeSavings   | Number            | Y         | EV driving saving during lifetime. |
+| gasCost           | Number            | Y         | Cost of gasoline used in savings calculations. |
+| fuelConsumption   | Number            | Y         | Miles per gallon used in savings calculations. |
+| ecost             | Number            | Y         | Cost of electricity from utility company. ($/kWh) |
```
```suggestion
| ecost             | Number            | Y         | Cost of electricity from utility company. (currency/kWh) |
```

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,215 @@
+| carBatterySize    | Number:Energy     | Y         | Car battery pack size. |
+| carBatteryRange   | Number:Length     | Y         | Car range. |
+| carChargingRate   | Number:Power      | Y         | Car charging rate. |
+
```
An example of manual configuration would be nice.

jsjames replied:
yes, will add when this works.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/JuiceNetBindingConstants.java:
```diff
@@ -0,0 +1,76 @@
+    public static final ThingTypeUID DEVICE_THING_TYPE = new ThingTypeUID(BINDING_ID, DEVICE);
+
+    // Device properties
+    public static final String DEVICE_PROP_NAME = "name";
```
This and other properties - proper prefix would increase readability:
```suggestion
    public static final String PROPERTY_NAME = "name";
```

jlaur reviewed (changes requested):
Last round - now almost done. I will need to take a closer look at the interactions between discovery and the bridge handler, but other than that I have now been through all files.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,230 @@
+        ContentResponse response;
+
+        params.put("cmd", "get_account_units");
+        params.put("device_id", bridgeUID.getAsString());
```
great suggestion.

jsjames commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,178 @@
+    @Override
+    public void dispose() {
+        logger.debug("Handler disposed.");
+        if (pollingJob != null) {
+            pollingJob.cancel(true);
```
makes sense - thanks for the education here.  I wasn't thinking about another thread and didn't understand why the check of the class variable didn't resolve the null pointer.

jsjames commented:
@jlaur, thanks again for all the constructive feedback.  I have updated a few areas that you will want to check.  I should have captured all the input.  The only one that I may not have fully grasped were the NPE.  Since I rewrote a lot of that, it would be helpful to provide any feedback.  Thanks.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,355 @@
+
+    private final Logger logger = LoggerFactory.getLogger(JuiceNetDeviceHandler.class);
+
+    protected final TimeZoneProvider timeZoneProvider;
+
+    // properties
+    protected String id = "";
+    protected String name = "";
+
+    protected String token = "";
+    protected long targetTimeTou = 0;
+    protected long lastInfoTimestamp = 0;
```
```suggestion
    private final TimeZoneProvider timeZoneProvider;

    // properties
    private String id = "";
    private String name = "";

    private String token = "";
    private long targetTimeTou = 0;
    private long lastInfoTimestamp = 0;
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,355 @@
+        this.token = token;
+
+        if (!name.equals(this.name)) {
+            Map<String, String> properties = editProperties();
+            properties.put(PROPERTY_NAME, name);
+            this.updateProperties(properties);
```
There is a convenience method for updating a single property:
```suggestion
            updateProperty(PROPERTY_NAME, name);
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,355 @@
+        String stringId = configuration.get(PARAMETER_UNIT_ID).toString();
+        if (stringId.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "Must include an id in the configuraiton for the device.");
```
Typo:
```suggestion
                    "Must include an id in the configuration for the device.");
```

But please also add this as i18n string in `juicenet.properties`, so it can be translated:
```java
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "@text/offline.configuration-error.id-missing");
```

Then in juicenet.properties:
```
offline.configuration-error.id-missing = Must include an id in the configuration for the device.
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,355 @@
+
+        if (token.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "Device does not exist as part of this JuiceNet Account");
```
Please add i18n string.

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
````diff
@@ -0,0 +1,227 @@
+
+Once, the JuiceNet Account thing has been created, all JuiceBox EV chargers associated with that account will be discovered and added to the inbox.
+
+## Manual Configuration Example
+
+The bridge and device can be manually configured in a .things file.
+Both the apiToken parameter for the Bridge and the unitID parameter for the EV charger can be found on the account page at https://home.juce.net.
+
+```
+Bridge juicenet:account:myaccount [ apiToken="xxxx-xxxx-xxxx-xxxx-xxxxx" ] {
+    Thing device JamesCharger [ unitID="xxxxxxx" ]
+}
+```
+
````
I would propose the following for being more complete and compliant with other other bindings:
- Move this chapter to be the last one, although probably before "Widget".
- Renamed chapter heading to "Full Example"
- Create sub chapters "juicenet.things" (with above), "juicenet.items" (new) etc.
- See for example https://www.openhab.org/addons/bindings/hue/ or other bindings.

jlaur replied:
I see you did some, I will post more specific comments for the few remaining suggestions. :)

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
```diff
@@ -0,0 +1,227 @@
+
+## Channels
+
+| channel           | type              | read-only | description                  |
```
Minor, but consider reformatting alignment of table, so it would also look pretty in source form.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,217 @@
+    private static final String API_ACCOUNT = API_HOST + "box_pin";
+    private static final String API_DEVICE = API_HOST + "box_api_secure";
+
+    protected String apiToken = "";
+    protected HttpClient httpClient;
+    protected ThingUID bridgeUID;
```
```suggestion
    private String apiToken = "";
    private HttpClient httpClient;
    private ThingUID bridgeUID;
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,217 @@
+        this.httpClient = httpClient;
+    }
+
+    public void initialize(String apiToken) {
```
Consider renaming method now that it does only one thing:
```suggestion
    public void setApiToken(String apiToken) {
```

This would be more straight-forward and readable.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/dto/JuiceNetApiDeviceChargingStatus.java:
```diff
@@ -0,0 +1,43 @@
+@NonNullByDefault
+public class JuiceNetApiDeviceChargingStatus {
+    @SerializedName("amps_limit")
+    public int currentLimit;
```
I know this is exhausting, but I would suggest to keep the API name here in the DTO and changing it only to comply with naming convention:
```suggestion
    public int ampsLimit;
```

Then DTO would be more or less 1:1 with API, and the abstraction would be in the handler which would translate from API names to channel names etc.

This applies to all DTO's.

I won't insist on this, but IMHO it would be cleaner to abstract things in the handler rather than directly in the DTO. Furthermore, you can't assume that each DTO member translates directly to a channel anyway. In some cases you might need some conversion or mapping. As an example, it could be relevant to convert miles to kilometers.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,204 @@
+        JuiceNetDiscoveryService discoveryService = this.discoveryService;
+
+        if (discoveryService == null) {
+            return;
```
Shouldn't this method still set online status and notify child handlers, even if discovery service is not (yet) set?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    private void pollDevices() {
+        logger.debug("pollDevices");
```
If needed, I would suggest **TRACE** level here, since it doesn't provide any information other than being called (Chinese debugging).

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,204 @@
+            JuiceNetDeviceHandler handler = (JuiceNetDeviceHandler) t.getHandler();
+            if (handler == null) {
+                logger.trace("no handler");
+                return;
```
```suggestion
                continue;
```

or is it intentional to skip other things?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,204 @@
+
+            JuiceNetDeviceHandler handler = (JuiceNetDeviceHandler) t.getHandler();
+            if (handler == null) {
+                logger.trace("no handler");
```
For this logging to be useful, probably some identification of thing should be included.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -0,0 +1,204 @@
+    public void initialize() {
+        config = getConfigAs(JuiceNetBridgeConfiguration.class);
+
+        logger.trace("JuiceNetBridgeHandler:initialize");
```
Name of class should be included automatically. It might be more useful with some identification of the bridge.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    public void goOnline() {
```
```suggestion
    private void goOnline() {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        // This device will go ONLINE on the first successful API call in queryDeviceStatusAndInfo
+    }
+
+    public void handleApiException(Exception e) {
```
```suggestion
    private void handleApiException(Exception e) {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    protected JuiceNetApi getApi() {
```
```suggestion
    private JuiceNetApi getApi() {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    protected JuiceNetApi getApi() {
+        Bridge bridge = Objects.requireNonNull(getBridge());
```
This will throw if bridge is not configured. This could be avoided by validation in `initialize`:

```java
        Bridge bridge = getBridge();
        if (bridge == null) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
                    "@text/offline.configuration-error.bridge-missing");
            return;
        }
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        logger.trace("queryStatusAndInfo");
+        ThingStatus status = getThing().getStatus();
+
+        if (status == ThingStatus.OFFLINE || status == ThingStatus.UNKNOWN) {
+            goOnline();
+            return;
+        }
+
+        if (this.getThing().getStatus() != ThingStatus.ONLINE) {
+            return;
+        }
```
For me this is hard to read, not sure if tricky or wrong. If offline/unknown, go online and return. If online, skip both conditionals. If other statuses, return. Is this some trickery for handling uninitialized, initializing etc. to not try anything?

jsjames replied:
yeah - this was leftover code.  It is redundant.

jlaur replied:
Still wondering if `status != ThingStatus.ONLINE` would be a safer condition here?

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    public void tryQueryDeviceStatusAndInfo() throws JuiceNetApiException, InterruptedException {
```
```suggestion
    private void tryQueryDeviceStatusAndInfo() throws JuiceNetApiException, InterruptedException {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    protected ZonedDateTime toZonedDateTime(long localEpochSeconds) {
```
```suggestion
    private ZonedDateTime toZonedDateTime(long localEpochSeconds) {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        return Instant.ofEpochSecond(localEpochSeconds).atZone(timeZoneProvider.getTimeZone());
+    }
+
+    protected void refreshStatusChannels() {
```
```suggestion
    private void refreshStatusChannels() {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java:
```diff
@@ -0,0 +1,349 @@
+        updateState(CHANNEL_CAR_CHARGING_RATE, new QuantityType<>(deviceCar.chargingRate, Units.WATT));
+    }
+
+    protected void refreshInfoChannels() {
```
```suggestion
    private void refreshInfoChannels() {
```

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/resources/OH-INF/thing/juicenet-device.xml:
```diff
@@ -0,0 +1,288 @@
+		</supported-bridge-type-refs>
+
+		<label>JuiceBox Charger</label>
+		<description>JuiceBox EV-Charger</description>
```
```suggestion
		<description>JuiceBox EV Charger</description>
```

jlaur reviewed (changes requested):
Almost at finish line - additional feedback after latest commits from re-reading changed classes.

jsjames commented:
This should have all the updates at this point!  Thanks again for all your time to review the code.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetBridgeHandler.java:
```diff
@@ -172,7 +168,10 @@ public void iterateApiDevices() {
             Thing childThing = getThingById(dev.unitId);
             if (childThing == null) {
-                discoveryService.notifyDiscoveryDevice(dev.unitId, dev.name);
+                JuiceNetDiscoveryService discoveryService = this.discoveryService;
```
This could be done before the loop as it only needs to be set once.

jlaur commented on bundles/org.openhab.binding.juicenet/README.md:
````diff
@@ -0,0 +1,266 @@
+
+A JuiceBox EV Charger requires a a unitID which can also be found in the device settings at the JuiceNet web page.
+
+If configuring the binding with manual configuration an example thing file looks like this:
+
+```
+Bridge juicenet:account:myaccount [ apiToken="xxxx-xxxx-xxxx-xxxx-xxxxx" ] {
+    Thing device JamesCharger [ unitID="xxxxxxx" ]
+}
+```
````
This should be moved below the **Channels** chapter and a header like "Example Things File" could be added. Or you could even have a small hierarchy:
- Level 2 header: Full Example
- Level 3 header: Things File
- Level 3 header: Items File

or similar.

jlaur commented on bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/api/JuiceNetApi.java:
```diff
@@ -0,0 +1,217 @@
+        try {
+            JsonObject jsonResponse = postApiCommand(ApiCommand.GET_ACCOUNT_UNITS, null);
+
+            listDevices = new Gson().fromJson(jsonResponse.get("units").getAsJsonArray(), JuiceNetApiDevice[].class);
```
`NullPointerException` possible - similar to https://github.com/openhab/openhab-addons/pull/10768#discussion_r1012297310.

jlaur reviewed (approved):
Thanks for your contribution.

Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

jlaur commented:
> Thanks for your contribution.
> 
> Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

@jsjames - I don't know if you saw this message, but it would be cool if you could create a PR adding a logo for your binding. :-)
